src/bufr/BufrParser/Query/File.cpp

```diff
@@ -13,96 +13,70 @@
 
 #include "QueryRunner.h"
 #include "QuerySet.h"
-#include "DataProvider.h"
+#include "DataProvider/DataProvider.h"
+#include "DataProvider/NcepDataProvider.h"
+#include "DataProvider/WmoDataProvider.h"
 
 
 namespace Ingester {
 namespace bufr {
-    File::File(const std::string &filename, bool isWmoFormat, const std::string &wmoTablePath) :
-            filename_(filename),
-            fileUnit_(nextFileUnit()),
-            fileUnitTable1_(nextFileUnit()),
-            fileUnitTable2_(nextFileUnit()),
-            isWmoFormat_(isWmoFormat),
-            wmoTablePath_(wmoTablePath)
+    File::File(const std::string &filename, bool isWmoFormat, const std::string &wmoTablePath)
     {
-        open();
-    }
-
-    void File::open()
-    {
-        open_f(fileUnit_, filename_.c_str());
-
-        if (!isWmoFormat_)
+        if (!isWmoFormat)
         {
-            openbf_f(fileUnit_, "IN", fileUnit_);
+            dataProvider_ = std::make_shared<Ingester::bufr::NcepDataProvider>(filename);
         }
         else
         {
-            openbf_f(fileUnit_, "SEC3", fileUnit_);
-
-            if (!wmoTablePath_.empty())
-            {
-                mtinfo_f(wmoTablePath_.c_str(), fileUnitTable1_, fileUnitTable2_);
-            }
+            dataProvider_ = std::make_shared<Ingester::bufr::WmoDataProvider>(filename,
+                                                                             wmoTablePath);
         }
+
+        dataProvider_->open();
     }
 
     void File::close()
     {
-        closbf_f(fileUnit_);
-        close_f(fileUnit_);
+        dataProvider_->close();
     }
 
-    void File::rewind() {
-        close();
-        open();
+    void File::rewind()
```

Comment:
Why do we want to rewind here?

Reply:
This just gives the user of the File class the ability to rewind the file (start over). It could be that we don't need a seperate DataProvider class.